### PR TITLE
ARROW-10070: [C++][Compute] Implement var and std aggregate kernel

### PR DESCRIPTION
**Pull request by cyb70289**

### CODE REVIEW

**Comment by cyb70289** [comment]
Dev / Lint ci failure looks not related.
```
INFO:archery:Running Docker linter
apache-rat license violation: go/arrow/flight/Flight_grpc.pb.go
apache-rat license violation: go/arrow/flight/example_flight_server_test.go
Error: `docker-compose --file /home/runner/work/arrow/arrow/docker-compose.yml run --rm ubuntu-lint` exited with a non-zero exit code 1, see the process log above.
```
missing license header?

**Comment by kszucs** [comment]
> missing license header?
Yes, https://github.com/apache/arrow/pull/8273 should fix that.

**Comment by jorisvandenbossche** [comment]
We might want to have an option to specify the denominator? (whether it is `n` or `n - 1`, to compute population vs sample standard deviation) As some examples, [numpy](https://numpy.org/doc/stable/reference/generated/numpy.std.html) has a `ddof` keyword, [postgres](https://www.postgresql.org/docs/9.1/functions-aggregate.html) or [clickhouse](https://clickhouse.tech/docs/en/sql-reference/aggregate-functions/reference/stddevpop/) have separate functions for both, [julia](https://docs.julialang.org/en/v1/stdlib/Statistics/#Statistics.std) has a `corrected` keyword.

**Comment by cyb70289** [comment]
> We might want to have an option to specify the denominator? (whether it is `n` or `n - 1`, to compute population vs sample standard deviation)

Thank you, will do.

**Comment by cyb70289** [comment]
Added option `ddof` to control stdev divisor. Same as [numpy.std](https://numpy.org/doc/stable/reference/generated/numpy.std.html).

**Comment by nealrichardson** [comment]
Two notes:

1. Naming: I've never seen this called `stdev` anywhere. `stddev` is common, in numpy and julia it's `std`, in R it's `sd`. Let's go with one of those. Maybe just add an extra "d"?
2. Since `sd = sqrt(var)` (https://github.com/apache/arrow/pull/8269/files#diff-461bd7e445c2a190f1173ebdefa21002R106), would it make sense to implement variance (i.e. most of this patch), and then standard deviation as the sqrt of that? That way we get two kernels (or even three, if sqrt is exposed as a kernel too).

**Comment by cyb70289** [comment]
Thanks @nealrichardson 

>     1. Naming: I've never seen this called `stdev` anywhere. `stddev` is common, in numpy and julia it's `std`, in R it's `sd`. Let's go with one of those. Maybe just add an extra "d"?

Naming is always the hardest thing :) Looks `std` is used more often, and it's short.
AFAIK, `stdev` is used in excel (the most popular statistic software I guess? :)

>     2. Since `sd = sqrt(var)` (https://github.com/apache/arrow/pull/8269/files#diff-461bd7e445c2a190f1173ebdefa21002R106), would it make sense to implement variance (i.e. most of this patch), and then standard deviation as the sqrt of that? That way we get two kernels (or even three, if sqrt is exposed as a kernel too).

I also thought about the `var` kernel. Will update this patch to include it.

**Comment by cyb70289** [comment]
Added variance kernel `var`. Renamed standard deviation kernel to `std`.

File: docs/source/cpp/compute.rst [hunk 1]
```diff
@@ -140,8 +140,12 @@ Aggregations
 +--------------------------+------------+--------------------+-----------------------+--------------------------------------------+
 | mode                     | Unary      | Numeric            | Scalar Struct  (2)    |                                            |
 +--------------------------+------------+--------------------+-----------------------+--------------------------------------------+
+| std                      | Unary      | Numeric            | Scalar Float64        | :struct:`VarStdOptions`                    |
++--------------------------+------------+--------------------+-----------------------+--------------------------------------------+
 | sum                      | Unary      | Numeric            | Scalar Numeric (3)    |                                            |
 +--------------------------+------------+--------------------+-----------------------+--------------------------------------------+
+| var                      | Unary      | Numeric            | Scalar Float64        | :struct:`VarStdOptions`                    |
```
Comment (pitrou) [inline docs/source/cpp/compute.rst:147]: Would rather "stddev" and "variance" respectively. "std" and "var" can easily be misunderstood, IMHO.

File: cpp/src/arrow/compute/kernels/aggregate_var_std.cc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    }
+    if (this->count == 0) {
+      this->count = state.count;
+      this->sum = state.sum;
+      this->m2 = state.m2;
+      return;
+    }
+    double delta = this->sum / this->count - state.sum / state.count;
+    this->m2 += state.m2 +
+                delta * delta * this->count * state.count / (this->count + state.count);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_var_std.cc:73]: Hmm... can you document the source for this formula? This doesn't seem to correspond:
https://www.emathzone.com/tutorials/basic-statistics/combined-variance.html

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 3]
```diff
@@ -914,5 +914,127 @@ TEST_F(TestInt32ModeKernel, LargeValueRange) {
...
+    reader.Next();
+  }
+  return m2 / count;
+}
+
+class TestVarStdKernelRandom : public TestPrimitiveVarStdKernel<DoubleType> {};
+
+TEST_F(TestVarStdKernelRandom, Basics) {
+  auto rand = random::RandomArrayGenerator(0x5487656);
+  auto array = rand.Numeric<DoubleType>(40000, -10000.0, 100000.0, 0.1);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1030]: Ironically, the fact that this is random identically-distributed data means that the slices will have similar means and variances.  Perhaps making the array smaller would make the test more significant.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 4]
```diff
@@ -914,5 +914,127 @@ TEST_F(TestInt32ModeKernel, LargeValueRange) {
...
+TYPED_TEST(TestNumericVarStdKernel, Basics) {
+  VarStdOptions options;  // ddof = 0, population var/std
+
+  this->AssertVarStdIs("[100]", options, 0);
+  this->AssertVarStdIs("[1, 2, 3]", options, 0.6666666666666666);
+  this->AssertVarStdIs("[null, 1, 2, null, 3]", options, 0.6666666666666666);
+
+  this->AssertVarStdIs({"[]", "[1]", "[2]", "[null]", "[3]"}, options,
+                       0.6666666666666666);
+  this->AssertVarStdIs({"[1, 2, 3]", "[4, 5, 6]", "[7, 8]"}, options, 5.25);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:984]: Can we also test with chunks of very different sizes? For example `[1, 2, 3, 4, 5, 6, 7]` and `[8]`.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 5]
```diff
@@ -914,5 +914,127 @@ TEST_F(TestInt32ModeKernel, LargeValueRange) {
...
+
+  this->AssertVarStdIsInvalid("[null, null, null]", options);
+  this->AssertVarStdIsInvalid("[]", options);
+
+  options.ddof = 1;  // sample var/std
+
+  this->AssertVarStdIs("[1, 2]", options, 0.5);
+  this->AssertVarStdIs({"[1, 2, 3]", "[4, 5, 6]", "[7, 8]"}, options, 6.0);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:992]: Same here.

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 6]
```diff
@@ -914,5 +914,127 @@ TEST_F(TestInt32ModeKernel, LargeValueRange) {
...
+
+TEST_F(TestVarStdKernelRandom, Basics) {
+  auto rand = random::RandomArrayGenerator(0x5487656);
+  auto array = rand.Numeric<DoubleType>(40000, -10000.0, 100000.0, 0.1);
+  auto chunked = *ChunkedArray::Make(
+      {array->Slice(0, 1000), array->Slice(1000, 9000), array->Slice(10000, 30000)});
+  double expected_var = WelfordVar(*array);
+
+  VarStdOptions options;
+  this->AssertVarStdIs(chunked, options, expected_var, 0.0001);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:1036]: Can we also test with `ddof = 1`? Though with such a large array size, the change may be rather small.

File: cpp/src/arrow/compute/kernels/aggregate_var_std.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:73]: The comment `// Combine variance from two chunks` is misleading.
Actually, this code is combining `m2` (which equals `sum((X-mean)^2)`) from two chunks, not the variance. See https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Parallel_algorithm
Will refine comments.

File: docs/source/cpp/compute.rst (hunk 1 above)
Comment (cyb70289) [reply docs/source/cpp/compute.rst]: done

File: cpp/src/arrow/compute/kernels/aggregate_var_std.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:74]: done

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc]: done

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1038]: done

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:1049]: done

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 6 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc]: done

File: cpp/src/arrow/compute/api_aggregate.h [hunk 7]
```diff
@@ -76,6 +76,18 @@ struct ARROW_EXPORT MinMaxOptions : public FunctionOptions {
   enum Mode null_handling = SKIP;
 };
 
+/// \brief Control Delta Degrees of Freedom (ddof) of Variance and Stddev kernel
+///
+/// The divisor used in calculations is N - ddof, where N is the number of elements.
+/// By default, ddof is zero, and population variance or stddev is returned.
+struct ARROW_EXPORT VarStdOptions : public FunctionOptions {
+  explicit VarStdOptions(int ddof = 0) : ddof(ddof) {}
```
Comment (arw2019) [inline cpp/src/arrow/compute/api_aggregate.h:84]: Naming suggestion: how about `DdofOptions`?

Comment (cyb70289) [reply cpp/src/arrow/compute/api_aggregate.h:84]: I'm not sure if we will add options later, e.g., to ignore NaN

Comment (jorisvandenbossche) [reply cpp/src/arrow/compute/api_aggregate.h:84]: I agree there might be other options in the future. But with the renaming of the functions, maybe call it `VarianceOptions` instead?

Comment (cyb70289) [reply cpp/src/arrow/compute/api_aggregate.h:84]: Kind of struggling as stddev kernel uses this same option.
Maybe export both `VarianceOptions` and `StddevOptions` and alias them as same type internally?

Comment (pitrou) [reply cpp/src/arrow/compute/api_aggregate.h:84]: Only one is useful. Just `VarianceOptions` IMHO.

File: cpp/src/arrow/compute/kernels/aggregate_var_std.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/aggregate_var_std.cc:74]: Thank you!

**Review by pitrou — APPROVED** [review]
+1, thank you @cyb70289

**Comment by jorisvandenbossche** [comment]
I suppose the behaviour with NaN is that any NaN in the input gives NaN as result? That might be worth adding a test for?

**Comment by pitrou** [comment]
Since we're not doing anything special, it certainly will. I think that can be tackled in a separate JIRA, when adding a nan handling option perhaps.

**Pull request closed without merging**